src/System.Net.Security/src/System/Net/SslSessionsCache.cs

```diff
@@ -178,23 +166,21 @@ internal static void CacheCredential(SafeFreeCredentials creds, byte[] thumbPrin
             {
                 if (GlobalLog.IsEnabled)
                 {
-                    GlobalLog.Print("CacheCredential() Refused to cache an Invalid Handle = " + creds.ToString() + ", Current Cache Count = " + s_CachedCreds.Count);
+                    GlobalLog.Print("CacheCredential() Refused to cache an Invalid Handle = " + creds.ToString() + ", Current Cache Count = " + s_cachedCreds.Count);
                 }
 
                 return;
             }
 
-            object key = new SslCredKey(thumbPrint, (int)sslProtocols, isServer, encryptionPolicy);
+            var key = new SslCredKey(thumbPrint, (int)sslProtocols, isServer, encryptionPolicy);
 
-            SafeCredentialReference cached = s_CachedCreds[key] as SafeCredentialReference;
+            SafeCredentialReference cached;
 
-            if (cached == null || cached.IsClosed || cached.Target.IsInvalid)
+            if (!s_cachedCreds.TryGetValue(key, out cached) || cached.IsClosed || cached.Target.IsInvalid)
             {
-                lock (s_CachedCreds)
+                lock (s_cachedCreds)
```

Comment:
Since we're using `ConcurrentDictionary` now, it may be possible to remove these locks, but I didn't have a chance to drill into it.